debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputReplicationMessage.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.postgresql.connection.pgoutput;
+
+import java.math.BigDecimal;
+import java.sql.SQLException;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.List;
+
+import org.apache.kafka.connect.errors.ConnectException;
+import org.postgresql.geometric.PGbox;
+import org.postgresql.geometric.PGcircle;
+import org.postgresql.geometric.PGline;
+import org.postgresql.geometric.PGlseg;
+import org.postgresql.geometric.PGpath;
+import org.postgresql.geometric.PGpoint;
+import org.postgresql.geometric.PGpolygon;
+import org.postgresql.jdbc.PgArray;
+import org.postgresql.util.PGInterval;
+import org.postgresql.util.PGmoney;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.connector.postgresql.PostgresType;
+import io.debezium.connector.postgresql.RecordsStreamProducer.PgConnectionSupplier;
+import io.debezium.connector.postgresql.connection.ReplicationMessage;
+import io.debezium.connector.postgresql.connection.wal2json.DateTimeFormat;
+import io.debezium.data.SpecialValueDecimal;
+import io.debezium.time.Conversions;
+import io.debezium.util.Strings;
+
+/**
+ * @author Gunnar Morling
+ * @author Chris Cranford
+ */
+public class PgOutputReplicationMessage implements ReplicationMessage {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PgOutputReplicationMessage.class);
+
+    private Operation op;
+    private Instant commitTimestamp;
+    private long transactionId;
+    private String table;
+    private List<Column> oldColumns;
+    private List<Column> newColumns;
+
+    public PgOutputReplicationMessage(Operation op, String table, Instant commitTimestamp, long transactionId, List<Column> oldColumns, List<Column> newColumns) {
+        this.op = op;
+        this.commitTimestamp = commitTimestamp;
+        this.transactionId = transactionId;
+        this.table = table;
+        this.oldColumns = oldColumns;
+        this.newColumns = newColumns;
+    }
+
+    @Override
+    public Operation getOperation() {
+        return op;
+    }
+
+    @Override
+    public Instant getCommitTime() {
+        return commitTimestamp;
+    }
+
+    @Override
+    public long getTransactionId() {
+        return transactionId;
+    }
+
+    @Override
+    public String getTable() {
+        return table;
+    }
+
+    @Override
+    public List<Column> getOldTupleList() {
+        return oldColumns;
+    }
+
+    @Override
+    public List<Column> getNewTupleList() {
+        return newColumns;
+    }
+
+    @Override
+    public boolean hasTypeMetadata() {
+        return true;
+    }
+
+    @Override
+    public boolean isLastEventForLsn() {
+        return true;
+    }
+
+    @Override
+    public boolean shouldSchemaBeSynchronized() {
```

Comment:
What will happen if you alter optionality of one of columns? Will it send a new `R` message? If not then this is a little bit risky and needs to be properly handled on the process side - simliar to PK update.

Reply:
@Naros Let me specify the question better, if the optionality is changed, is it a reason to send the `R` message not immideatly but with the next update? I just want to make sure that the `R` message will arrive at some point of time.

Reply:
Any form of schema change on a table will eventually trigger a `R` event; however, we only will get that event if and only if an insert, update, or delete operation occurs after the table was altered.  If no rows are ever added, changed, or removed after the alter occurred, then we'll never see an `R` event for that table.

Reply:
ok, thanks, so if optionality change also triggers the `R` message at some point of time then this is fine for me.